forte/mrdsrg-spin-integrated/dsrg_mrpt2_grad/dsrg_mrpt2_deriv_write_rdms.cc

```diff
@@ -0,0 +1,553 @@
+    temp["xYuV"] += 0.25 * Gamma2_["uVxY"];
+
+    // CI contribution
+    temp.block("aaaa")("xyuv") += 0.5 * 0.25 * cc2aa_bra("Iuvxy") * x_ci("I");
+    temp.block("AAAA")("XYUV") += 0.5 * 0.25 * cc2bb_bra("IUVXY") * x_ci("I");
+    temp.block("aAaA")("xYuV") += 0.5 * 0.25 * cc2ab_bra("IuVxY") * x_ci("I");
+
+    temp.block("aaaa")("xyuv") += 0.5 * 0.25 * cc2aa_ket("Juvxy") * x_ci("J");
+    temp.block("AAAA")("XYUV") += 0.5 * 0.25 * cc2bb_ket("JUVXY") * x_ci("J");
+    temp.block("aAaA")("xYuV") += 0.5 * 0.25 * cc2ab_ket("JuVxY") * x_ci("J");
+
+    // all-alpha and all-beta
+    temp2["ckdl"] += temp["cdkl"];
+    temp2["cldk"] -= temp["cdkl"];
+    // alpha-beta
+    temp2["ckDL"] += 2.0 * temp["cDkL"];
+    temp2["clDK"] += 2.0 * temp["cDlK"];
+    temp.zero();
+
+    temp["eumv"] += 2.0 * Z["em"] * Gamma1_["uv"];
+    temp["EUMV"] += 2.0 * Z["EM"] * Gamma1_["UV"];
+    temp["eUmV"] += 2.0 * Z["em"] * Gamma1_["UV"];
+
+    temp["u,a1,n,u1"] += 2.0 * Z["un"] * Gamma1_["u1,a1"];
+    temp["U,A1,N,U1"] += 2.0 * Z["UN"] * Gamma1_["U1,A1"];
+    temp["u,A1,n,U1"] += 2.0 * Z["un"] * Gamma1_["U1,A1"];
+
+    temp["v,a1,u,u1"] += Z["uv"] * Gamma1_["u1,a1"];
+    temp["V,A1,U,U1"] += Z["UV"] * Gamma1_["U1,A1"];
+    temp["v,A1,u,U1"] += Z["uv"] * Gamma1_["U1,A1"];
+
+    // <[F, T2]>
+    if (X5_TERM || X7_TERM) {
+        temp["aviu"] += Sigma3["ia"] * Gamma1_["uv"];
+        temp["AVIU"] += Sigma3["IA"] * Gamma1_["UV"];
+        temp["aViU"] += Sigma3["ia"] * Gamma1_["UV"];
+    }
+    // <[V, T1]>
+    if (X6_TERM || X7_TERM) {
+        temp["aviu"] += Xi3["ia"] * Gamma1_["uv"];
+        temp["AVIU"] += Xi3["IA"] * Gamma1_["UV"];
+        temp["aViU"] += Xi3["ia"] * Gamma1_["UV"];
+    }
+
+    // all-alpha and all-beta
+    temp2["ckdl"] += temp["cdkl"];
+    temp2["cldk"] -= temp["cdkl"];
+    // alpha-beta
+    temp2["ckDL"] += 2.0 * temp["cDkL"];
+
+    temp2.iterate(
+        [&](const std::vector<size_t>& i, const std::vector<SpinType>& spin, double& value) {
+            if (std::fabs(value) > 1e-12) {
+                if (spin[2] == AlphaSpin) {
+                    d2aa.write_value(i[0], i[1], i[2], i[3], 0.5 * value, 0, "outfile", 0);
+                    d2bb.write_value(i[0], i[1], i[2], i[3], 0.5 * value, 0, "outfile", 0);
+                } else {
+                    d2ab.write_value(i[0], i[1], i[2], i[3], value, 0, "outfile", 0);
+                }
+            }
+        });
+
+    d2aa.flush(1);
+    d2bb.flush(1);
+    d2ab.flush(1);
+
+    d2aa.set_keep_flag(1);
+    d2bb.set_keep_flag(1);
+    d2ab.set_keep_flag(1);
+
+    d2aa.close();
+    d2bb.close();
+    d2ab.close();
+
+    outfile->Printf("Done");
+}
+
+void DSRG_MRPT2::tpdm_backtransform() {
+    // Backtransform the TPDM
+
+    std::vector<std::shared_ptr<psi::MOSpace>> spaces;
+    spaces.push_back(psi::MOSpace::all);
+    std::shared_ptr<TPDMBackTransform> transform =
+        std::shared_ptr<TPDMBackTransform>(new TPDMBackTransform(
```

Comment:
It's better practice to use `make_shared` here. It saves a little bit of computer time, rather than make a raw pointer with new and then cast to `shared_ptr`.